src/Spanner/Session/CacheSessionPool.php

```diff
@@ -0,0 +1,548 @@
+            $item = $this->cacheItemPool->getItem($this->cacheKey);
+            $data = $item->get();
+
+            unset($data['inUse'][$session->name()]);
+            array_push($data['queue'], [
+                'name' => $session->name(),
+                'expiration' => $session->expiration()
+                    ?: $this->time() + SessionPoolInterface::SESSION_EXPIRATION_SECONDS
+            ]);
+            $this->cacheItemPool->save($item->set($data));
+        });
+    }
+
+    /**
+     * Clear the session pool.
+     */
+    public function clear()
+    {
+        $this->cacheItemPool->clear();
+    }
+
+    /**
+     * Set the database used to make calls to manage sessions.
+     *
+     * @param Database $database The database.
+     */
+    public function setDatabase(Database $database)
+    {
+        $this->database = $database;
+        $identity = $database->identity();
+        $this->cacheKey = sprintf(self::CACHE_KEY_TEMPLATE, $identity['instance'], $identity['database']);
+    }
+
+    /**
+     * Get the underlying cache implementation.
+     *
+     * @return CacheItemPoolInterface
+     */
+    public function cacheItemPool()
+    {
+        return $this->cacheItemPool;
+    }
+
+    /**
+     * Get the current unix timestamp.
+     *
+     * @return int
+     */
+    protected function time()
+    {
+        return time();
+    }
+
+    /**
+     * Builds out a list of timestamps indicating the start time of the intent
+     * to create a session.
+     *
+     * @param array $data
+     * @param bool $hasSession
+     * @return array
+     */
+    private function buildToCreateList(array $data, $hasSession)
+    {
+        $number = 0;
+        $toCreate = [];
+        $time = $this->time();
+        $count = $this->getSessionCount($data);
+
+        if ($count < $this->config['minSessions']) {
+            $number = $this->config['minSessions'] - $count;
+        } elseif (!$hasSession && !$data['queue'] && $count < $this->config['maxSessions']) {
+            $number++;
+        }
+
+        for ($i = 0; $i < $number; $i++) {
+            $toCreate[uniqid($time . '_')] = $time;
+        }
+
+        return $toCreate;
+    }
+
+    /**
+     * Purge any items in the to create queue that have been inactive for 20
+     * minutes or more.
+     *
+     * @param array $data
+     */
+    private function purgeOrphanedToCreateItems(array &$data)
+    {
+        foreach ($data['toCreate'] as $key => $timestamp) {
+            if ($timestamp + 1200 < $this->time()) {
+                unset($data['toCreate'][$key]);
+            }
+        }
+    }
+
+    /**
+     * Purge any in use sessions that have been inactive for 20 minutes or more.
+     *
+     * @param array $data
+     */
+    private function purgeOrphanedInUseSessions(array &$data)
+    {
+        foreach ($data['inUse'] as $key => $session) {
+            if ($session['lastActive'] + 1200 < $this->time()) {
+                unset($data['inUse'][$key]);
```

Comment:
It should be noted this code is only executed if we have met the session pool max limit. The concern is when a session gets checked out, it is possible for a process to stop and never attempt to return the session to the queue. Instead of 20 minutes, what if we defaulted to 80?
That way we would have a good idea that the session is most likely expired at that point. The issue there, is if a streaming operation takes 80 minutes in a single process. Is that a common operation we expect to occur?